File: app/src/main/java/com/alexharman/stitchathon/KnitPackage/KnitPattern.java
package com.alexharman.stitchathon.KnitPackage;

import java.util.ArrayList;

public class KnitPattern {

    public Stitch[][] stitches;
    public String name;

    // Takes width of stitches into account
    private int currentDistanceInRow = 0;
    private int totalStitchesDone = 0;
    private int totalStitches = 0;
    private int currentRow = 0;
    private int nextStitchInRow = 0;
    private int patternWidth = 0;
    private int rows = 0;
    public ArrayList<Stitch> stitchTypes = new ArrayList<>();

    // Assuming doubleknit for now. Will be false for knitting on the round
    private boolean oddRowsOpposite = true;

    public KnitPattern(String[][] pattern, String name) {
        this(pattern);
        this.name = name;
    }

    public KnitPattern(ArrayList<ArrayList<String>> pattern, String name) {
        this(pattern);
        this.name = name;
    }

    KnitPattern(String[][] pattern) {
        buildPattern(pattern);
    }

    KnitPattern(ArrayList<ArrayList<String>> pattern) {
        String[][] arPattern = new String[pattern.size()][0];
        for (int i = 0; i < pattern.size(); i++) {
            arPattern[i] = pattern.get(i).toArray(new String[pattern.get(i).size()]);
        }
        buildPattern(arPattern);
    }

    private void buildPattern(String[][] pattern) {
        int rowWidth;
        ArrayList<String> stitchTypeStrings = new ArrayList<>();
        stitches = new Stitch[pattern.length][];
        for (int row = 0; row < pattern.length; row++) {
            rowWidth = 0;
            stitches[row] = new Stitch[pattern[row].length];
            for (int col = 0; col < pattern[row].length; col++) {
                addNewStitchType(stitchTypeStrings, pattern[row][col]);
                totalStitches++;
                stitches[row][col] = new Stitch(pattern[row][col]);
                rowWidth += stitches[row][col].getWidth();
            }
            patternWidth = Math.max(patternWidth, rowWidth);
        }
        rows = pattern.length;

        for (String stitchString: stitchTypeStrings) {
            stitchTypes.add(new Stitch(stitchString));
        }
    }

    private void addNewStitchType(ArrayList<String> stitchTypes, String stitch) {
        if (!stitchTypes.contains(stitch)) {
            if (stitch.contains("/")) {
                for (String s : stitch.split("/")) {
                    addNewStitchType(stitchTypes, s);
                }
            }
            stitchTypes.add(stitch);
        }
    }

    public void increment() {
        if (currentRow == stitches.length-1 && (nextStitchInRow < 0 || nextStitchInRow > stitches[stitches.length-1].length)) {
            return;
        }

        currentDistanceInRow += stitches[currentRow][nextStitchInRow].getWidth();
        totalStitchesDone++;
        stitches[currentRow][nextStitchInRow].setDone(true);
        nextStitchInRow += getRowDirection();
        if (isEndOfRow()) {
            currentRow++;
            currentDistanceInRow = 0;
            nextStitchInRow = getStartOfRow();
        }
    }

    public int incrementRow() {
        if (currentRow == stitches.length-1 && (nextStitchInRow < 0 || nextStitchInRow > stitches[stitches.length-1].length)) {
            return 0;
        }
        int newStitchesDone = 0;
        int direction = getRowDirection();

        for (int i = nextStitchInRow; i*direction <= getEndOfRow(); i += direction) {
            stitches[currentRow][i].setDone(true);
            newStitchesDone++;
        }

        totalStitchesDone += newStitchesDone;
        currentRow++;
        nextStitchInRow = getStartOfRow();
        currentDistanceInRow = 0;
        return newStitchesDone;
    }

    public void undoStitch() {
        if(currentRow == 0 && nextStitchInRow == getStartOfRow()) {
            return;
        }

        if (isStartOfRow()) {
            currentRow--;
            nextStitchInRow = getEndOfRow();
            currentDistanceInRow = 0;
            for (int c = Math.min(getStartOfRow(), getEndOfRow()); c < Math.max(getStartOfRow(), getEndOfRow()); c += 1) {
                currentDistanceInRow += stitches[currentRow][c].getWidth();
            }
        } else {
            currentDistanceInRow -= stitches[currentRow][nextStitchInRow].getWidth();
            nextStitchInRow -= getRowDirection();
        }
        stitches[currentRow][nextStitchInRow].setDone(false);
        totalStitchesDone--;
    }

    public int getTotalStitchesDone() {
        return totalStitchesDone;
    }

    public int getTotalStitches() {
        return totalStitches;
    }

    public int getCurrentRow() {
        return currentRow;
    }

    public int getNextStitchInRow() {
        return nextStitchInRow;
    }

    public int getRowDirection() {
        return oddRowsOpposite && (currentRow % 2 == 1) ? -1 : 1;
    }

    public boolean isEndOfRow() {
        if (getRowDirection() == 1) {
            return nextStitchInRow == stitches[currentRow].length;
        }
        return nextStitchInRow == -1;
    }

    public boolean isStartOfRow() {
        if (getRowDirection() == 1) {
            return nextStitchInRow == 0;
        }
        return nextStitchInRow == stitches[currentRow].length - 1;
    }

    private int getEndOfRow() {
        if (getRowDirection() == 1) {
            return stitches[currentRow].length - 1;
        }
        return 0;
    }

    private int getStartOfRow() {
        if (getRowDirection() == 1) {
            return 0;
        }
        return stitches[currentRow].length - 1;
    }

    public int getStitchesLeftInRow() {
        if (getRowDirection() == 1) {
            return stitches[currentRow].length - nextStitchInRow;
        }
        return nextStitchInRow + 1;
    }

    public int getStitchesDoneInRow() {
        if (getRowDirection() == 1) {
            return nextStitchInRow;
        }
        return stitches[currentRow].length - nextStitchInRow - 1;
    }

    public int getPatternWidth() {
        return patternWidth;
    }

    public int getRows() {
        return rows;
    }

    public int getCurrentDistanceInRow() {
        return currentDistanceInRow;
    }
}

File: app/src/main/java/com/alexharman/stitchathon/KnitPackage/KnitPattern.kt
package com.alexharman.stitchathon.KnitPackage

class KnitPattern(
        val name: String,
        val stitches: Array<Array<Stitch>>,
        val oddRowsOpposite: Boolean = true,
        currentRow: Int = 0,
        stitchesDoneInRow: Int = 0) {

    val stitchTypes: Set<Stitch>
    val totalStitches: Int
    val patternWidth: Int

    // Takes width of stitches into account
    var currentDistanceInRow = 0
        private set
    var stitchesDoneInRow = 0
        private set
    var totalStitchesDone = 0
        private set
    var currentRow = 0
        private set

    init {
        this.currentRow = currentRow
        this.stitchesDoneInRow = stitchesDoneInRow
        totalStitches = findTotalStitches(stitches)
        patternWidth = findPatternWidth(stitches)
        totalStitchesDone = findTotalStitchesDone(stitches, currentRow)
        currentDistanceInRow = findCurrentDistanceInRow(stitches, currentRow)

        val initStitchTypes = stitches.flatten().toSet()
        stitchTypes = initStitchTypes.union(initStitchTypes.flatMap { it.madeOf.toList() }.toSet())
    }

    constructor(name: String = "", pattern: Array<Array<String>>, oddRowsOpposite: Boolean = true) :
            this (name, pattern.map { row -> row.map { Stitch(it) }.toTypedArray() }.toTypedArray(), oddRowsOpposite)

    constructor(name: String = "", pattern: List<List<String>>, oddRowsOpposite: Boolean = true) :
            this(name, Array(pattern.size, { i -> pattern[i].toTypedArray() }), oddRowsOpposite)

    fun increment (numStitches: Int = 1) {
        for (i in 0 until numStitches) {
            if (isFinished) return

            currentDistanceInRow += stitches[currentRow][nextStitchInRow].width
            totalStitchesDone++
            stitchesDoneInRow++

            if (!isFinished && isEndOfRow) {
                currentRow++
                currentDistanceInRow = 0
                stitchesDoneInRow = 0
            }
        }
    }

    fun incrementRow() = increment(stitchesLeftInRow)

    fun undoStitch() {
        if (currentRow == 0 && isStartOfRow)
            return

        if (isStartOfRow) {
            currentRow--
            stitchesDoneInRow = stitches[currentRow].size - 1
            currentDistanceInRow = stitches[currentRow].sumBy { it.width } - stitches[currentRow][nextStitchInRow].width
        } else {
            stitchesDoneInRow--
            currentDistanceInRow -= stitches[currentRow][nextStitchInRow].width
        }
        totalStitchesDone--
    }

    private fun findTotalStitches(stitches: Array<Array<Stitch>>) = stitches.sumBy { it.size }

    private fun findPatternWidth(stitches: Array<Array<Stitch>>) =
            stitches.maxBy { it.sumBy { it.width } }?.sumBy { it.width } ?: 0

    private fun findTotalStitchesDone(stitches: Array<Array<Stitch>>, currentRow: Int) =
            stitchesDoneInRow + stitches.sliceArray(IntRange(0, currentRow - 1)).sumBy { it.size }

    private fun findCurrentDistanceInRow(stitches: Array<Array<Stitch>>, currentRow: Int): Int {
        val rowDoneRange =
                if(currentRowDirection == 1)
                    IntRange(0, nextStitchInRow - 1)
                else
                    IntRange(nextStitchInRow + 1, stitches[currentRow].size-1)
        return stitches[currentRow].sliceArray(rowDoneRange).sumBy { it.width }
    }

    fun rowDirection(row: Int = currentRow) =
            if (oddRowsOpposite && (row % 2 == 1)) -1 else 1

    val currentRowDirection: Int
        get() = rowDirection(currentRow)

    private val isEndOfRow: Boolean
        get() = stitchesLeftInRow == 0

    val isStartOfRow: Boolean
        get() = stitchesDoneInRow == 0

    val stitchesLeftInRow: Int
        get() = stitches[currentRow].size - stitchesDoneInRow

    val nextStitchInRow: Int
        get() = if (currentRowDirection == 1) stitchesDoneInRow else (stitches[currentRow].size - 1 - stitchesDoneInRow)

    val numRows: Int
        get() = stitches.size

    val isFinished: Boolean
        get() = totalStitches == totalStitchesDone
}
